docs/proposals/local-registries.md

```diff
@@ -0,0 +1,165 @@
+[Rancher](https://github.com/rancher/k3d/blob/master/docs/registries.md#using-a-local-registry),
+[microk8s](https://microk8s.io/docs/registry-built-in), and
+[OpenShift](https://docs.openshift.com/container-platform/4.5/registry/architecture-component-imageregistry.html).
+This proposal will provide a means of declaring that an image spec references a local registry,
+thereby allowing host information to be removed from the image reference.
+
+### Goals
+
+- Simplify how local image registries can be referenced in `Build` objects.
+- Provide a means of declaring that an image references a cluster-local image registry.
+
+### Non-Goals
+
+- Automate how pull/push secrets can be added to a `Build` object.
+- Support secure image pull/push if a local registry uses self-signed certificates.
+
+## Proposal
+
+Application developers will be able to specify `local: true|false` if an image reference points
+to a cluster-local image registry.
+
+The build controller/operator will watch the global "LocalRegistryHosting" `ConfigMap` to check if
+a local image registry is declared. If declared, it will infer the local registry host information
+from what is provided in the `ConfigMap`.
+
+When a `BuildRun` is instantiated, the image specs that declare `local: true` are mutated to inject
+the local image registry host.
+
+### User Stories [optional]
+
+As an application developer
+I would like to declare that images in builds reference a local image registry
+So that I do not have to provide host information in the pull spec
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Obtaining local registry information
+
+First, clusters which deploy a local image registry must provide the releavant host info by
+publishing it in the `local-registry-hosting`
+[ConfigMap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap).
+The build operator/controller must have RBAC which allows it to read the `ConfigMap` and obtain the
+registry host information. For `Build` and `BuildRun` objects, the `HostFromClusterNetwork` value
+should be used as the hostname.
+
+If the `local-registry-hosting` ConfigMap is created, updated, or deleted, the build controller
+and/or operator needs to react accordingly.
+
+#### Declaring an image as local
+
+`Image` references in the API will be updated to have the `local` field. An example using the 
+`output` spec:
+
+```yaml
+spec:
+  output:
+    image: mynamespace/myapp:latest
+    local: true
+```
+
+When an image is declared as `local`, the build controller prepends the local image registry
+hostname to the provided image reference.
+
+### Risks and Mitigations
+
+App developers use `local: true` on clusters that don't have a local registry defined, end up
+pulling images from `docker.io` or an image registry in container runtime's unqualified search
+path.
```

Comment:
I would consider a build invalid that defines `local: true` in a cluster that has no local registry. Its registered status should be false therefore.

Reply:
+1

Reply:
+1, and I am thinking do we need have the `local` parameter in buildrun? I think maybe the parameter in build is enough.

Reply:
I think we want it in `BuildRun` too. IMO we should allow raw `BuildRun` objects to be created without an associated `Build`, and still take advantage of this feature.